### PR TITLE
First AKS integration draft

### PR DESCRIPTION
**Pull request by lovew-lc**
This is a first draft for the AKS integration in hopsworks.ai. For now I have documented one example setup, other setups (e.g. public AKS) are in large parts the same as the example. I'll add more examples if we gain enough requests for it.

### CODE REVIEW

File: docs/hopsworksai/azure/aks_acr_integration.md [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+2. One Azure container registry configured in your account
+3. One AKS cluster
+4. Permissions to the ACR and AKS attached to a user-managed identity
+
+This guide on example setup with a private AKS cluster and public ACR.
+!!! note
+A public AKS cluster means the kubernetes API server is accessible outside the virtual network it is deployed in. Similarly, a public ACR is accessible through the internet.
+
+
+## User-managed identity
```
Comment (jimdowling) [inline docs/hopsworksai/azure/aks_acr_integration.md:17]: User assigned managed identity (managed identity)

File: docs/hopsworksai/azure/aks_acr_integration.md [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+
+!!! note
+The managed identity is attached to the virtual machines that runs inside your subscription. Hence, the permissions only applies to services that runs within your subscription. 
+
+AKS and ACR integration requires some additional permissions to be attached to the managed identity used by the hopsworks cluster. 
+
+To setup the managed identity, go to managed identities in the azure portal and then click add. Place the managed indentity in the same resource group as your are planning to create the hopsworks cluster in. In addition, choose a name for the managed identity, then review and create.
```
Comment (jimdowling) [inline docs/hopsworksai/azure/aks_acr_integration.md:24]: .., go to the resource group where you will add the managed identity - this should be the same resource group as you deploy Hopsworks in. In your selected resource group, click on the Access Control (IAM) menu item, and then click on add button. In the search dialog, enter "user assigned managed identity", and then select a name for the managed identity, review it and create it.

File: docs/hopsworksai/azure/aks_acr_integration.md [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+
+Next go to the created managed identity in order to create role assignments. Go to azure role assignments and add the following roles with scope of the resource group you are planning to use:
```
Comment (jimdowling) [inline docs/hopsworksai/azure/aks_acr_integration.md:26]: This is adding a managed identity role at the subscription level, which is fine. But we should also have text to describe how to add the role assignment to a resource group.

###Add role assignment at subscription level
.....
###Add role assignment at resource group level
Go to the resource group where you created your managed identity. Select the Access Control (IAM) menu item, then click on "Add", then "Add role Assignment". Select the roles "AcrPull", "AcrPush", "Azure Kubernetes Service Contributor Role" and then select the managed identity that you are assignment them to.

File: docs/hopsworksai/azure/aks_acr_integration.md [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+    <figcaption>AKS permissions</figcaption>
+  </figure>
+</p>
+
+## Private AKS cluster and public ACR
+
+This guide will step through setting up a private AKS cluster and a public ACR. 
+
+### Step 1 Create an AKS cluster
+Go to kubernetes services in the azure portal and click add. Place the kubernetes cluster in the same resource group as the hopsworks cluster will reside in and choose a name for the kubernetes cluster.
```
Comment (jimdowling) [inline docs/hopsworksai/azure/aks_acr_integration.md:51]: (will reside in) is not needed here.

Comment (jimdowling) [reply docs/hopsworksai/azure/aks_acr_integration.md:51]: you could write
the same resource group that the Hopsworks cluster is located in, and choose a...

File: docs/hopsworksai/azure/aks_acr_integration.md [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+    <a  href="../../../assets/images/hopsworksai/azure/aks-base.png">
+      <img src="../../../assets/images/hopsworksai/azure/aks-base.png" alt="AKS general configuration">
+    </a>
+    <figcaption>AKS general configuration</figcaption>
+  </figure>
+</p>
+
+Next click the Authentication tab and verify the settings are identical to the picture below.
+!!! note
+Currently AKS is only supported through managed identities.
```
Comment (jimdowling) [inline docs/hopsworksai/azure/aks_acr_integration.md:64]: Contact the Logical Clocks sales team if you have a self-managed Kubernetes cluster.

File: docs/hopsworksai/azure/aks_acr_integration.md [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+  </figure>
+</p>
+
+Next press review and create, then create the cluster.
+
+Because the kubernetes API service is private the hopsworks cluster must be able to reach it over a private network. There are two options to integrate with a private AKS cluster. The first option is to put the hopsworks cluster in a pre-defined virtual network with peering setup to the kubernetes network. The second option is to create a subnet inside the kubernetes virtual network where the hopsworks cluster will be placed.
+
+#### Step 1 option a Peering setup
+
+In order to create a peering first a virtual network for the hopsworks cluster must be setup. Go to virtual networks and press add.
```
Comment (jimdowling) [inline docs/hopsworksai/azure/aks_acr_integration.md:114]: In order to establish virtual peering between the kubernetes cluster and Hopsworks, you need to select or create a virtual network for Hopsworks.

Note: Hopsworks, not hopsworks. This mistake is in many places in the text.

File: docs/hopsworksai/azure/aks_acr_integration.md [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+<p align="center">
+  <figure>
+    <a  href="../../../assets/images/hopsworksai/azure/aks-peering1.png">
+      <img src="../../../assets/images/hopsworksai/azure/aks-peering1.png" alt="Virtual network peering">
+    </a>
+    <figcaption>Virtual network peering configuration</figcaption>
+  </figure>
+</p>
+
+For the virtual network select the virtual network which was created by AKS, in our example this was hopsworksstagevnet154. Then press add.
```
Comment (jimdowling) [inline docs/hopsworksai/azure/aks_acr_integration.md:143]: For the virtual network, select ... that was created by AKS. In our example, this was ...

File: docs/hopsworksai/azure/aks_acr_integration.md [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+<p align="center">
+  <figure>
+    <a  href="../../../assets/images/hopsworksai/azure/aks-vnet-link-config.png">
+      <img src="../../../assets/images/hopsworksai/azure/aks-vnet-link-config.png" alt="">
+    </a>
+    <figcaption>Private DNS link configuration</figcaption>
+  </figure>
+</p>
+
+The setup is now finalized and you can create the hopsworks cluster.
```
Comment (jimdowling) [inline docs/hopsworksai/azure/aks_acr_integration.md:187]: Hopsworks

File: docs/hopsworksai/azure/aks_acr_integration.md [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+
+#### Step 1 option b Subnet in AKS network
+
+With this setup the hopsworks cluster will reside in the same virtual network as the AKS cluster. The difference is that a new subnet in the virtual network will be used for the hopsworks cluster.
+
+To setup the subnet, first go to the virtual network that was created by AKS, in our example this was hopsworksstagevnet154. Next go to the subnets tab.
```
Comment (jimdowling) [inline docs/hopsworksai/azure/aks_acr_integration.md:193]: . In our example, this was ....

File: docs/hopsworksai/azure/aks_acr_integration.md [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+<p align="center">
+  <figure>
+    <a  href="../../../assets/images/hopsworksai/azure/aks-subnet.png">
+      <img src="../../../assets/images/hopsworksai/azure/aks-subnet.png" alt="">
+    </a>
+    <figcaption>AKS subnet setup</figcaption>
+  </figure>
+</p>
+
+Press + Subnet. Choose a name for the subnet, for example "hopsworks" and an IP range that does not overlap with the kubernetes network. Then save.
```
Comment (jimdowling) [inline docs/hopsworksai/azure/aks_acr_integration.md:204]: . for example, "hopsworks", and an IP

File: docs/hopsworksai/azure/aks_acr_integration.md [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+<p align="center">
+  <figure>
+    <a  href="../../../assets/images/hopsworksai/azure/aks-subnet-config.png">
+      <img src="../../../assets/images/hopsworksai/azure/aks-subnet-config.png" alt="">
+    </a>
+    <figcaption>AKS subnet setup</figcaption>
+  </figure>
+</p>
+
+### Step 2 Create hopsworks cluster
```
Comment (jimdowling) [inline docs/hopsworksai/azure/aks_acr_integration.md:215]: Create the Hopsworks cluster

File: docs/hopsworksai/azure/aks_acr_integration.md [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+
+This step assumes you create a cluster in hopsworks.ai The AKS configuration can be set under the Managed containers tab. Select the Use Azure AKS and Azure ACR as enabled. Two new fields with the name of the container registry you created and the AKS cluster name will pop up. In the previous example we created an ACR with name hopsaksregistry and an AKS cluster with name hopsaks-cluster. Hence, the configuration should look similar to the picture below
```
Comment (jimdowling) [inline docs/hopsworksai/azure/aks_acr_integration.md:217]: assumes you are creating your Hopsworks cluster using hopsworks.ai
or
you have created your Hopsworks cluster using hopsworks.ai

**Review by jimdowling — CHANGES_REQUESTED** [review]
A few minor changes. The only bigger one is that you should describe how to create the "user-assigned managed identity" at the resource group level - not just at the subscription level. Both levels are possible, so we should describe both approaches.

File: docs/hopsworksai/azure/aks_acr_integration.md (hunk 3 above)
Comment (lovew-lc) [reply docs/hopsworksai/azure/aks_acr_integration.md:26]: I think you got this one the other way around. In the example the roles are applied at resource group level. The scope is chosen when you attach the role. Anyhow, I'll clarify this in the text.

File: docs/hopsworksai/azure/aks_acr_integration.md [hunk 13]
```diff
@@ -0,0 +1,229 @@
+# Integration with Azure AKS and ACR
+
+This guide shows how to create a cluster in hopsworks.ai with integrated support for Azure Kubernetes Service (AKS) and Azure Container Registry (ACR). This enables Hopsworks to launch Python jobs, Jupyter servers, and ML model servings on top of AKS.
```
[the remark this answers is not in the document]
Comment (lovew-lc) [reply docs/hopsworksai/azure/aks_acr_integration.md]: fixed

File: docs/hopsworksai/azure/aks_acr_integration.md [hunk 14]
```diff
@@ -0,0 +1,229 @@
...
+
+Hopsworks AKS and ACR integration have four requirements:
+
+1. A virutal network with access to AKS pods and the AKS API servers 
+2. One Azure container registry configured in your account
+3. One AKS cluster
+4. Permissions to the ACR and AKS attached to a user-managed identity
+
+This guide on example setup with a private AKS cluster and public ACR.
```
[the remark this answers is not in the document]
Comment (lovew-lc) [reply docs/hopsworksai/azure/aks_acr_integration.md]: Yes, updated in the comment

File: docs/hopsworksai/azure/aks_acr_integration.md [hunk 15]
```diff
@@ -0,0 +1,229 @@
...
+!!! note
+A public AKS cluster means the kubernetes API server is accessible outside the virtual network it is deployed in. Similarly, a public ACR is accessible through the internet.
+
+
+## User-managed identity
+
+!!! note
+The managed identity is attached to the virtual machines that runs inside your subscription. Hence, the permissions only applies to services that runs within your subscription.
```
[the remark this answers is not in the document]
Comment (lovew-lc) [reply docs/hopsworksai/azure/aks_acr_integration.md]: Fixed

File: docs/hopsworksai/azure/aks_acr_integration.md (hunk 1 above)
Comment (lovew-lc) [reply docs/hopsworksai/azure/aks_acr_integration.md]: Fixed

File: docs/hopsworksai/azure/aks_acr_integration.md (hunk 2 above)
Comment (lovew-lc) [reply docs/hopsworksai/azure/aks_acr_integration.md]: Fixed

File: docs/hopsworksai/azure/aks_acr_integration.md (hunk 3 above)
Comment (lovew-lc) [reply docs/hopsworksai/azure/aks_acr_integration.md]: Reformulated it a bit in the latest commit.

File: docs/hopsworksai/azure/aks_acr_integration.md (hunk 4 above)
Comment (lovew-lc) [reply docs/hopsworksai/azure/aks_acr_integration.md]: Fixed

File: docs/hopsworksai/azure/aks_acr_integration.md (hunk 5 above)
Comment (lovew-lc) [reply docs/hopsworksai/azure/aks_acr_integration.md]: Added to the note

File: docs/hopsworksai/azure/aks_acr_integration.md (hunk 6 above)
Comment (lovew-lc) [reply docs/hopsworksai/azure/aks_acr_integration.md]: Muscle memory from coding. I have capitalized Hopsworks everywhere in the file.

**Review by jimdowling — APPROVED** [review]

**Pull request merged**